contracts/contracts/strategies/balancer/BalancerMetaPoolStrategy.sol

```diff
@@ -0,0 +1,477 @@
+            address poolAsset = address(tokens[i]);
+            // Convert the balancer pool asset to the vault collateral asset
+            address asset = fromPoolAsset(poolAsset);
+            // Get the balancer pool assets withdraw from the pool plus any that were already in this strategy contract
+            uint256 poolAssetAmount = IERC20(poolAsset).balanceOf(
+                address(this)
+            );
+
+            // Unwrap assets like wstETH and sfrxETH to rebasing assets stETH and frxETH
+            uint256 assetAmount = 0;
+            if (poolAssetAmount > 0) {
+                assetAmount = unwrapPoolAsset(asset, poolAssetAmount);
+            }
+
+            // Transfer the vault collateral assets to the vault
+            if (assetAmount > 0) {
+                IERC20(asset).safeTransfer(vaultAddress, assetAmount);
+                emit Withdrawal(asset, platformAddress, assetAmount);
+            }
+        }
+    }
+
+    /**
+     * @notice Approves the Balancer pool to transfer all supported
+     * assets from this strategy.
+     * Also approve any suppered assets that are wrapped in the Balancer pool
+     * like stETH and frxETH, to be transferred from this strategy to their
+     * respective wrapper contracts. eg wstETH and sfrxETH.
+     *
+     * Is only executable by the Governor.
+     */
+    function safeApproveAllTokens()
+        external
+        override
+        onlyGovernor
+        nonReentrant
+    {
+        uint256 assetCount = assetsMapped.length;
+        for (uint256 i = 0; i < assetCount; ++i) {
+            _approveAsset(assetsMapped[i]);
+        }
+        _approveBase();
+    }
+
+    // solhin t-disable-next-line no-unused-vars
+    function _abstractSetPToken(address _asset, address) internal override {
+        address poolAsset = toPoolAsset(_asset);
+        if (_asset == stETH) {
+            IERC20(stETH).safeApprove(wstETH, 1e50);
+        } else if (_asset == frxETH) {
+            IERC20(frxETH).safeApprove(sfrxETH, 1e50);
+        }
+        _approveAsset(poolAsset);
+    }
+
+    /**
+     * @dev Approves the Balancer Vault to transfer an asset from
+     * this strategy. The assets could be a Vault collateral asset
+     * like WETH or rETH; or a Balancer pool asset that wraps the vault asset
+     * like wstETH or sfrxETH.
+     */
+    function _approveAsset(address _asset) internal {
+        IERC20 asset = IERC20(_asset);
+        asset.safeApprove(address(balancerVault), 0);
+        asset.safeApprove(address(balancerVault), type(uint256).max);
+    }
+
+    /**
+     * @notice Returns the rate supplied by the Balancer configured rate
+     * provider. Rate is used to normalize the token to common underlying
+     * pool denominator. (ETH for ETH Liquid staking derivatives)
+     *
+     * @param _asset Address of the Balancer pool asset
+     * @return rate of the corresponding asset
+     */
+    function getRateProviderRate(address _asset)
```

Comment:
The rate providers are hard coded, correct? Can we just cache a mapping when creating the contract? Would much rather make one external call than three here, given that balance checking is an area we care about gas on.

Reply:
It is immutable with the MetaStable pool we are fork testing with. I'd like to understand the composable stable pools better because they have a different implementation of it. We are not supporting those yet and looking at [the implementation ](https://github.com/balancer/balancer-v2-monorepo/blob/master/pkg/pool-stable/contracts/ComposableStablePoolRates.sol)they seem to honor the and keep indexes the same.

I completely agree with you that this change should be made for the reasons already outlined. I would just introduce it more carefully with more time to research that we are completely safe.